[READY] Adds selector lists to SDQL

JJRcop opened the pull request:
## About The Pull Request

I've added a selector array query. It's a special construct formed like this:
```bnf
'@[' <selectors> ']'
```
The `<selectors>` is the very same way that you normally use to select datums for all of the queries.
```bnf
UPDATE <selectors> SET <assignments>
SELECT <selectors>
```

I chose `@[ ]` because it's adding an @ sign to the existing SDQL list symbols, since it still generates a list.

## Why It's Good For The Game

It adds more flexibility to SDQL, allowing admins to do more with the language. I find it very convenient in testing to really narrow down the search to say, an /area, so SDQL will only look for datums in that /area's .contents, when paired with MAP.

![query](https://user-images.githubusercontent.com/5211576/56059908-573f6b00-5d33-11e9-94c5-c423a5e9cbb4.png)
![result](https://user-images.githubusercontent.com/5211576/56059913-59092e80-5d33-11e9-9c93-d1acff71f870.png)

## Changelog
:cl: JJRcop
add: SDQL2 now features a list constructor that takes a selector query. 
tweak: (sub expressions) will work with var access, like .vars and [list indexes]
/:cl:

i stayed up too late

JJRcop committed:
[SDQL2] Adds selector arrays

JJRcop committed:
SDQL2 documentation spec updated slightly

JJRcop committed:
SDQL2 selector array docs + tweaks

If a single list is the only result of a MAP statement,
meaning objs.len == 1, AND it's a list, it will unwrap
the result from its containing list and return it directly.

The selector array logic keeps track of the state of the query
when it starts the first Search. It won't change the state back
to Execute until the top-level is reached, which sets a var
to keep track.

optimumtact commented:
dude just revive NTSL and give it the same hooks for admins

it's a lot easier to think about imperatively

JJRcop committed:
SDQL2 MAP change from objs.len to length(objs)

JJRcop commented:
I'm not very interested in reviving NTSL right now. While it would be much better than SDQL, I would have to do so much more work than this, including research on making/maintaining a programming language and its various systems, along with the specific language attributes NTSL has.

It might not be as hard as I'm thinking it is, but I don't have the desire right now.

AnturK commented on code/modules/admin/verbs/SDQL2/SDQL_2.dm:
```diff
@@ -535,6 +539,8 @@ GLOBAL_DATUM_INIT(sdql2_vv_statobj, /obj/effect/statclick/SDQL2_VV_all, new(null
 					objs[j] = SDQL_expression(x, expression)
 					SDQL2_TICK_CHECK
 					SDQL2_HALT_CHECK
+				if(length(objs) == 1 && islist(objs[1]))
```
What's with conditional unwrapping, this is going to be an issue if anyone wants to retrieve list variables. This should probably be controlled by the user.

JJRcop replied:
This is to get around an issue where defining a list then immediately accessing it doesn't work. I'll see what I can do about reverting this and fixing the issue itself.

JJRcop replied:
I've reverted the unwrapping and allowed variable to process expressions now. The proper solution to my above problem is `(@[/area/crew_quarters/bar MAP contents])[1]`

JJRcop committed:
[SDQL2] Reverts MAP list unwrap

JJRcop committed:
[SDQL] Move (expr) from expr, unary_expr to var

This should result in the same functionality but instead of expressions and unary_expressions processing sub expressions, variables do
This means that expression->value->variable is now completely recursable

JJRcop committed:
[SDQL2] Update selector list documentation

JJRcop commented:
Resolved @AnturK's concerns.
